test/topology_custom/test_read_repair.py

```diff
@@ -0,0 +1,127 @@
+                                                                "INSERT INTO ks.tbl (pk, v) VALUES ({}, 1)",
+                                                                "SELECT * FROM ks.tbl",
+                                                                {"pk": 0, "v": 1},
+                                                                {"pk": 1000, "v": 1},
+                                                                id="partition-tombstone")
+incremental_repair_test_data = [incremental_repair_row_tombstone_test_data,
+                                incremental_repair_partition_tombstone_test_data]
+
+
+@pytest.mark.parametrize("columns_spec,delete_query,insert_query,select_query,row0,row1000", incremental_repair_test_data)
+@pytest.mark.asyncio
+async def test_incremental_read_repair(columns_spec, delete_query, insert_query, select_query, row0, row1000, manager: ManagerClient):
+    """Stress the incremental read repair logic
+
+    Write a long stream of row tombstones, with a live row before and after.
+    """
+    cmdline = ["--hinted-handoff-enabled", "0", "--query-tombstone-page-limit", "1000"]
+    node1 = await manager.server_add(cmdline=cmdline)
+    node2 = await manager.server_add(cmdline=cmdline)
+
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
+
+    cql = manager.get_cql()
+
+    cql.execute("CREATE KEYSPACE ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 2}")
+    cql.execute(f"CREATE TABLE ks.tbl ({columns_spec}) WITH speculative_retry = 'NONE'")
+
+    delete_stmt = cql.prepare(delete_query)
+    for k in range(0, 20000):
+        cql.execute(delete_stmt, (k,))
+
+    # Only on node1
+    async with stopped_node(manager, node2):
+        logger.info("Writing row0 to node1")
+        cql.execute(SimpleStatement(insert_query.format(0), consistency_level=ConsistencyLevel.ONE))
+
+    # Only on node2
+    async with stopped_node(manager, node1):
+        logger.info("Writing row1000 to node2")
+        cql.execute(SimpleStatement(insert_query.format(1000), consistency_level=ConsistencyLevel.ONE))
+
+    def check_rows(res, expected_rows):
+        res_rows = []
+        nr_pages = 0
+        while True:
+            res_rows.extend(list(res.current_rows))
+            nr_pages += 1
+            if res.has_more_pages:
+                res.fetch_next_page()
+            else:
+                break
+
+        logger.debug(f"check_rows(): read {nr_pages} pages")
+        assert nr_pages > 1
+
+        assert len(res_rows) == len(expected_rows)
+        for row, expected_row in zip(res_rows, expected_rows):
+            for k, v in expected_row.items():
+                assert getattr(row, k) == v
+
+    logger.info("Check rows with CL=ONE before read-repair")
+    stmt = SimpleStatement(select_query, consistency_level=ConsistencyLevel.ONE)
+    async with connection(manager, node1) as cql:
```

Comment:
Is having those separate connections necessary?

We had to do a lot of fine tuning of Python driver settings for the connection that you obtain through `manager.get_cql()`, in order to fix flakiness. Without doing the same fine tuning for these new connections this test is very likely going to be flaky (Python driver timeouts and so on).

But I wonder if you need the separate connections at all. Do you want to execute a query on a particular node? Establishing a new connection like this does not guarantee which node we'll execute query on (the list passed to `Cluster` is only for initial discovery, the driver automatically establishes connections to each node when discovering them.)

In order to query a particular node, you can use `manager.get_cql()` but pass `host` parameter to `execute` or `run_async`. The `host` parameter expects a `Host` object which you can obtain using `wait_for_cql_and_get_hosts`. See e.g. test_cluster_features.py how it uses `get_supported_features`.

Reply:
In my new version I updated this to:
```
 18 @contextlib.asynccontextmanager
 19 async def exclusive_connection(manager, node):
 20     try:
 21         profile = ExecutionProfile(request_timeout=30, load_balancing_policy = WhiteListRoundRobinPolicy([node.ip_addr]))
 22         c = Cluster([node.ip_addr], execution_profiles={EXEC_PROFILE_DEFAULT: profile})
 23         s = c.connect()
 24         yield s
 25     finally:
 26         s.shutdown()
```

`manager_client`'s API to obtain new connections is very clunky so I ended up avoiding it as much as I could.

Reply:
I need an exclusive connection to a single node, to verify that each has only the data it's supposed to have, without having to stop the other node (which is slow).

Reply:
> manager_client's API to obtain new connections is very clunky so I ended up avoiding it as much as I could.

My point is, you don't need new connections. You can use the single connection you already have (i.e. the one given by `manager.get_cql()`)

> I need an exclusive connection to a single node, to verify that each has only the data it's supposed to have, without having to stop the other node (which is slow).

You don't need an exclusive connection for that -- you can

Reply:
> In my new version I updated this to:

Look into test/topology/conftest.py, `cluster_con` function, all the timeout related parameters there and comments.
You should do the same adjustments.

But:
- you can use `cluster_con` and then the `whitelist` profile when doing queries so it sends the query to the node you initially specified as a contact point
- OR as I would prefer -- use the connection that `manager` already has, passing `host` parameter when executing queries.

Reply:
Well actually neither works.
* Using a new exclusive connection doesn't work for the partition-tombstone test-case: range scans on distribued tables can still decide to contact other nodes, there is nothing the client can  do to prevent that.
* Stopping one node and reading from the other also doesn't work, for some reason, some of the time rows are missing from the result.
* Passing `host=node_ip` to `session.execute()` just doesn't work at all. It always fails with `NoHostAvailable`, even if I reconnect the driver one line before trying it.

I have one more approach to try before I give up and just go back to dtests. This is just incredibly painful and frustrting.

Reply:
> As I explained in https://github.com/scylladb/scylladb/pull/14923#discussion_r1315637682, you need to pass an object of the driver's Host type to the host param. You can translate ServerInfos (node1 or node2) to Hosts using wait_for_cql_and_get_hosts.

Example from test_fencing.py:
```python3
    host2 = (await wait_for_cql_and_get_hosts(manager.cql, [servers[2]], time.time() + 60))[0]

    logger.info(f"trying to write through host2 to regular column [{host2}]")
    with pytest.raises(WriteFailure, match="stale topology exception"):
        await manager.cql.run_async("insert into t1(pk, int_c) values (1, 1)", host=host2)
```

Reply:
The driver silently accepted an ip address too and then said the node is down.

But anyway I am not trying a new approach, which doesn't rely on stopping/starting nodes and trying to control which node a query executes on. The approaches above (exclusive query, passing something to host) only work for single-page/non-paged queries, as soon as you have hundreds of pages, scylla can and does decide to send requests to other replicas as well. Especially if range-scans are involved.

Reply:
> But how would moving to dtests solve these problems? They're using the exact same Python driver.

Yes, writing tests which test repair are just frustrating, regardless of the framework. I expected it to be less painful if I move away from dtest and it is not. I said I will move back to dtest because the test was reliable working there already, it has become flaky after the porting. But now I remembered that I also added change after porting and those would have possibly made the test flaky with dtests too.

Reply:
> > The driver silently accepted an ip address too and then said the node is down.
> 
> Yup. That's the Python driver for you. :(

But didn't we decide that typeless languages are great the other day?

Reply:
> > The driver silently accepted an ip address too and then said the node is down.
> 
> Yup. That's the Python driver for you. :(

Well, I saw it in the docs that I should pass a `cassandra.Pool.Host` or something, but like a good Eastern European I thought "surely this speed-limit only applies to cyclists" and passed an ip anyway.

Reply:
> > > The driver silently accepted an ip address too and then said the node is down.
> > 
> > 
> > Yup. That's the Python driver for you. :(
> 
> But didn't we decide that typeless languages are great the other day?

Don't spread the flames. :)) :fire: